### ISSUE TITLE
Shared memory de-allocation problem

### ISSUE THREAD

**Issue by qiranq99** [v6d-io/v6d#1630]
Hi contributors, 

Here is a use case:
```
store = ShmStore() # It encapsulates vineyard
ipc_socket = store.open() # basically it starts a vineyard server in a subprocess
data = np.random.rand(1000,1000,1000) # ~7.4 GB
cli = vineyard.connect(ipc_socket)
cli.put(data)

# cli.close() # close client connection

store.close() # It terminates the vineyard server process (a subprocess)
time.sleep(10) # wait for subprocess termination
skt = store.open() # open a server again, getting a different socket
cli = vineyard.connect(skt)
cli.put(data)

```

Theoretically, after the server is closed or terminated, I would expect `shm` resource to be released immediately to serve other usage (under the circumstance that the terminated server is the last remaining instance), i.e., in the above case the `shm` usage would be exhibiting a transition pattern of `0 -> 7 -> 0 -> 7 -> 0 GB`. 

However, it is `0->7->14->0` in reality, indicating `shm` is not released after terminating the server. However, if the client is closed before the termination of the server, as commented in the codes, `shm` would be released on server termination. 
Logically, we would like `vineyard` to release all resource it allocated on the termination of the last remaining server regardless of the existence of the clients. 

Besides, an API like `vineyard_instance.flush()` may be desired to allow the users to flush the `shm` space occupied by the previously cached but later deleted objects (memory fragments), instead of letting the `shm` usage increases or stays at peak while the utilization is sometimes extremely low. 

Basically, 2 issues are discussed here:
1. shared memory de-allocation on the shutdown of the last server;
2. a (Python) API exposed to the users to enable explicit shared memory management. 

Cheers :()

**Comment by sighingnow** [comment]
> However, it is 0->7->14->0 in reality, indicating shm is not released after terminating the server. 

Does "terminating the server" mean killing the vineyardd process?

> Besides, an API like vineyard_instance.flush() may be desired to allow the users to flush the shm space occupied by the previously cached but later deleted objects (memory fragments)

Allocator will cache the physical memory and doesn't return it back to OS immediately after `free()`, see also https://github.com/v6d-io/v6d/issues/1620

**Comment by sighingnow** [comment]
I'll try verify the case provided in description as soon as possible.

**Comment by sighingnow** [comment]
A new API called `memory_trim` is added to vineyard client in #1638 to force release the physical memory in unused shared memory regions immediately.

See also the test case added in #1638: `test_memory_trim()`.

Hope that could be helpful for your questions.

**Comment by sighingnow** [comment]
The memory releasing can be verified by observing the system-wide memory usage like `free -h`.

**Comment by qiranq99** [comment]
Hi @sighingnow , the fix seems to be promising! Do you have a plan for a new release recently? As building from source is tough in our production environment. Thanks a lot!

**Comment by sighingnow** [comment]
A new release will be cut in tomorrow.

**Comment by sighingnow** [comment]
> A new release will be cut in tomorrow.

/cc @qiranq99, a new release (v0.19.1) has been uploaded to pypi.
